File: igris-overture/sdk/javascript/src/escapevector/thompson-router.ts
/**
 * Thompson Sampling Router - Local fallback with Bayesian optimization
 */

import fetch from 'node-fetch';
import { BanditArm, BayesianState, BetaSampler } from './bayesian-state';

export interface InferRequest {
  model: string;
  messages: Array<{ role: string; content: string }>;
  max_tokens?: number;
  temperature?: number;
  top_p?: number;
}

export interface InferResponse {
  id?: string;
  object?: string;
  created?: number;
  model: string;
  choices: Array<{
    index: number;
    message: { role: string; content: string };
    finish_reason?: string;
  }>;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

/**
 * Circuit Breaker - tracks provider health with atomic operations
 */
class CircuitBreaker {
  private failures = 0;
  private isOpen = false;
  private lastOpened = 0;

  constructor(private threshold: number) {}

  recordSuccess(): void {
    this.failures = 0;
    this.isOpen = false;
  }

  recordFailure(): void {
    this.failures++;
    if (this.failures >= this.threshold) {
      this.isOpen = true;
      this.lastOpened = Date.now();
    }
  }

  getIsOpen(): boolean {
    // Auto-reset after 30 seconds
    if (this.isOpen && Date.now() - this.lastOpened > 30000) {
      this.isOpen = false;
      this.failures = 0;
    }
    return this.isOpen;
  }
}

/**
 * Thompson Sampling Router
 */
export class ThompsonRouter {
  private circuitBreakers: Map<string, CircuitBreaker> = new Map();
  private abortControllers: Map<string, AbortController> = new Map();

  constructor(private state: BayesianState) {
    // Initialize circuit breakers
    for (const arm of state.arms) {
      this.circuitBreakers.set(
        arm.providerId,
        new CircuitBreaker(state.circuitBreakerThreshold)
      );
    }
  }

  /**
   * Perform inference using Thompson Sampling
   */
  async infer(req: InferRequest): Promise<InferResponse> {
    if (this.state.speculativeExecution) {
      return this.speculativeInfer(req);
    }

    // Standard Thompson Sampling with retries
    const maxRetries = this.state.maxRetries || 2;
    let lastError: Error | null = null;

    for (let retry = 0; retry <= maxRetries; retry++) {
      const arm = this.selectArmThompsonSampling();
      if (!arm) {
        throw new Error('No healthy providers available');
      }

      try {
        const startTime = Date.now();
        const response = await this.inferWithArm(arm, req);
        const latencyMs = Date.now() - startTime;

        // Update Bayesian parameters on success
        this.updateArmSuccess(arm, latencyMs);
        this.circuitBreakers.get(arm.providerId)!.recordSuccess();

        return response;
      } catch (error) {
        const latencyMs = Date.now() - Date.now();
        this.updateArmFailure(arm, latencyMs);
        this.circuitBreakers.get(arm.providerId)!.recordFailure();
        lastError = error as Error;

        // Exponential backoff
        if (retry < maxRetries) {
          await new Promise((resolve) =>
            setTimeout(resolve, 100 * Math.pow(2, retry))
          );
        }
      }
    }

    throw new Error(
      `All providers failed after ${maxRetries} retries: ${lastError?.message}`
    );
  }

  /**
   * Thompson Sampling selection
   */
  private selectArmThompsonSampling(): BanditArm | null {
    const healthyArms = this.getHealthyArms();
    if (healthyArms.length === 0) {
      return null;
    }

    // Exploration with probability ε
    if (Math.random() < this.state.explorationRate) {
      return healthyArms[Math.floor(Math.random() * healthyArms.length)];
    }

    // Exploitation: Sample from Beta(α, β) for each arm
    let bestArm: BanditArm | null = null;
    let maxSample = -1;

    for (const arm of healthyArms) {
      const sample = BetaSampler.sample(arm.alpha, arm.beta);
      if (sample > maxSample) {
        maxSample = sample;
        bestArm = arm;
      }
    }

    return bestArm;
  }

  /**
   * Speculative execution - race multiple providers
   */
  private async speculativeInfer(req: InferRequest): Promise<InferResponse> {
    const arms = this.getTopArmsByMean(3);
    if (arms.length === 0) {
      throw new Error('No healthy providers for speculative execution');
    }

    // Create abort controllers for each request
    const controllers = arms.map(() => new AbortController());

    // Launch parallel requests
    const promises = arms.map((arm, index) =>
      this.inferWithArm(arm, req, controllers[index].signal).then(
        (response) => ({
          response,
          armId: arm.providerId,
          latencyMs: Date.now(),
          error: null,
        }),
        (error) => ({
          response: null,
          armId: arm.providerId,
          latencyMs: Date.now(),
          error,
        })
      )
    );

    // Wait for first successful response
    const results = await Promise.allSettled(promises);

    for (let i = 0; i < results.length; i++) {
      if (results[i].status === 'fulfilled') {
        const result = (results[i] as PromiseFulfilledResult<any>).value;
        if (result.response) {
          // Cancel other requests
          controllers.forEach((c, idx) => {
            if (idx !== i) c.abort();
          });

          // Update winner
          const arm = arms[i];
          this.updateArmSuccess(arm, 0);
          this.circuitBreakers.get(arm.providerId)!.recordSuccess();

          return result.response;
        }
      }
    }

    throw new Error('All speculative requests failed');
  }

  /**
   * Make inference request to specific arm
   */
  private async inferWithArm(
    arm: BanditArm,
    req: InferRequest,
    signal?: AbortSignal
  ): Promise<InferResponse> {
    const endpoint = `${arm.endpoint}/chat/completions`;

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(arm.apiKey ? { Authorization: `Bearer ${arm.apiKey}` } : {}),
      },
      body: JSON.stringify(req),
      signal,
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`Provider returned error ${response.status}: ${text}`);
    }

    return (await response.json()) as InferResponse;
  }

  /**
   * Update arm on success
   */
  private updateArmSuccess(arm: BanditArm, latencyMs: number): void {
    arm.alpha += 1.0;
    arm.totalSelections++;
    arm.totalSuccesses++;

    // Update composite rewards
    const latencyScore = this.normalizeLatency(latencyMs);
    arm.avgLatencyScore = this.updateMovingAverage(
      arm.avgLatencyScore,
      latencyScore,
      arm.totalSelections
    );
    arm.avgSuccessRate = arm.totalSuccesses / arm.totalSelections;
    arm.updatedAt = Date.now();
  }

  /**
   * Update arm on failure
   */
  private updateArmFailure(arm: BanditArm, latencyMs: number): void {
    arm.beta += 1.0;
    arm.totalSelections++;
    arm.totalFailures++;

    const latencyScore = this.normalizeLatency(latencyMs);
    arm.avgLatencyScore = this.updateMovingAverage(
      arm.avgLatencyScore,
      latencyScore,
      arm.totalSelections
    );
    arm.avgSuccessRate = arm.totalSuccesses / arm.totalSelections;
    arm.updatedAt = Date.now();
  }

  /**
   * Normalize latency to 0-1 score
   */
  private normalizeLatency(latencyMs: number): number {
    if (latencyMs <= 0) return 1.0;
    const score = 1.0 - latencyMs / this.state.maxLatencyMs;
    return Math.max(0, Math.min(1, score));
  }

  /**
   * Exponential moving average
   */
  private updateMovingAverage(
    current: number,
    newValue: number,
    count: number
  ): number {
    if (count === 1) return newValue;
    const alpha = 2.0 / (count + 1);
    return alpha * newValue + (1 - alpha) * current;
  }

  /**
   * Get healthy arms (circuit breakers closed)
   */
  private getHealthyArms(): BanditArm[] {
    return this.state.arms.filter(
      (arm) => !this.circuitBreakers.get(arm.providerId)?.getIsOpen()
    );
  }

  /**
   * Get top N arms by expected value
   */
  private getTopArmsByMean(n: number): BanditArm[] {
    const healthy = this.getHealthyArms();
    healthy.sort((a, b) => {
      const meanA = a.alpha / (a.alpha + a.beta);
      const meanB = b.alpha / (b.alpha + b.beta);
      return meanB - meanA;
    });
    return healthy.slice(0, n);
  }

  /**
   * Update state (called when control plane returns)
   */
  updateState(state: BayesianState): void {
    this.state = state;

    // Update circuit breakers for new arms
    for (const arm of state.arms) {
      if (!this.circuitBreakers.has(arm.providerId)) {
        this.circuitBreakers.set(
          arm.providerId,
          new CircuitBreaker(state.circuitBreakerThreshold)
        );
      }
    }
  }

  /**
   * Get current state
   */
  getState(): BayesianState {
    return this.state;
  }
}

File: igris-overture/sdk/python/schlep/escapevector/thompson_router.py
"""Thompson Sampling Router - Local fallback with Bayesian optimization"""

import time
import requests
from typing import Dict, List, Optional, Any
from dataclasses import dataclass

from .bayesian_state import BanditArm, BayesianState


@dataclass
class InferRequest:
    """Inference request"""
    model: str
    messages: List[Dict[str, str]]
    max_tokens: Optional[int] = None
    temperature: Optional[float] = None
    top_p: Optional[float] = None


@dataclass
class InferResponse:
    """Inference response"""
    id: Optional[str] = None
    object: Optional[str] = None
    created: Optional[int] = None
    model: str = ""
    choices: List[Dict[str, Any]] = None
    usage: Optional[Dict[str, int]] = None

    def __post_init__(self):
        if self.choices is None:
            self.choices = []


class CircuitBreaker:
    """Circuit breaker for provider health tracking"""

    def __init__(self, threshold: int):
        self.threshold = threshold
        self.failures = 0
        self.is_open = False
        self.last_opened = 0

    def record_success(self):
        """Reset circuit breaker on success"""
        self.failures = 0
        self.is_open = False

    def record_failure(self):
        """Increment failure count"""
        self.failures += 1
        if self.failures >= self.threshold:
            self.is_open = True
            self.last_opened = time.time()

    def get_is_open(self) -> bool:
        """Check if circuit is open (auto-reset after 30s)"""
        if self.is_open and time.time() - self.last_opened > 30:
            self.is_open = False
            self.failures = 0
        return self.is_open


class ThompsonRouter:
    """Thompson Sampling Router"""

    def __init__(self, state: BayesianState):
        self.state = state
        self.circuit_breakers: Dict[str, CircuitBreaker] = {}

        # Initialize circuit breakers
        for arm in state.arms:
            self.circuit_breakers[arm.provider_id] = CircuitBreaker(
                state.circuit_breaker_threshold
            )

        # Session for HTTP requests
        self.session = requests.Session()
        self.session.headers.update({'Content-Type': 'application/json'})

    def infer(self, req: InferRequest) -> InferResponse:
        """Perform inference using Thompson Sampling"""
        if self.state.speculative_execution:
            return self._speculative_infer(req)

        # Standard Thompson Sampling with retries
        max_retries = self.state.max_retries or 2
        last_error = None

        for retry in range(max_retries + 1):
            arm = self._select_arm_thompson_sampling()
            if not arm:
                raise RuntimeError("No healthy providers available")

            try:
                start_time = time.time()
                response = self._infer_with_arm(arm, req)
                latency_ms = int((time.time() - start_time) * 1000)

                # Update Bayesian parameters on success
                self._update_arm_success(arm, latency_ms)
                self.circuit_breakers[arm.provider_id].record_success()

                return response

            except Exception as error:
                latency_ms = 0
                self._update_arm_failure(arm, latency_ms)
                self.circuit_breakers[arm.provider_id].record_failure()
                last_error = error

                # Exponential backoff
                if retry < max_retries:
                    time.sleep(0.1 * (2 ** retry))

        raise RuntimeError(
            f"All providers failed after {max_retries} retries: {last_error}"
        )

    def _select_arm_thompson_sampling(self) -> Optional[BanditArm]:
        """Thompson Sampling selection"""
        healthy_arms = self._get_healthy_arms()
        if not healthy_arms:
            return None

        # Exploration with probability ε
        import random
        if random.random() < self.state.exploration_rate:
            return random.choice(healthy_arms)

        # Exploitation: Sample from Beta(α, β) for each arm
        best_arm = None
        max_sample = -1.0

        for arm in healthy_arms:
            sample = arm.sample_beta()
            if sample > max_sample:
                max_sample = sample
                best_arm = arm

        return best_arm

    def _speculative_infer(self, req: InferRequest) -> InferResponse:
        """Speculative execution - race multiple providers"""
        # Get top 3 arms by expected value
        arms = self._get_top_arms_by_mean(3)
        if not arms:
            raise RuntimeError("No healthy providers for speculative execution")

        # Launch parallel requests (simplified - not truly parallel in Python)
        results = []
        for arm in arms:
            try:
                start_time = time.time()
                response = self._infer_with_arm(arm, req)
                latency_ms = int((time.time() - start_time) * 1000)
                results.append((response, arm, latency_ms, None))
                break  # Return first success
            except Exception as error:
                results.append((None, arm, 0, error))

        # Return first successful response
        for response, arm, latency_ms, error in results:
            if response:
                self._update_arm_success(arm, latency_ms)
                self.circuit_breakers[arm.provider_id].record_success()
                return response

        raise RuntimeError("All speculative requests failed")

    def _infer_with_arm(self, arm: BanditArm, req: InferRequest) -> InferResponse:
        """Make inference request to specific arm"""
        endpoint = f"{arm.endpoint}/chat/completions"

        headers = {}
        if arm.api_key:
            headers['Authorization'] = f"Bearer {arm.api_key}"

        payload = {
            "model": req.model,
            "messages": req.messages,
        }
        if req.max_tokens:
            payload["max_tokens"] = req.max_tokens
        if req.temperature:
            payload["temperature"] = req.temperature
        if req.top_p:
            payload["top_p"] = req.top_p

        response = self.session.post(
            endpoint,
            json=payload,
            headers=headers,
            timeout=self.state.timeout_ms / 1000,
        )

        if response.status_code >= 400:
            raise RuntimeError(
                f"Provider returned error {response.status_code}: {response.text}"
            )

        data = response.json()
        return InferResponse(**data)

    def _update_arm_success(self, arm: BanditArm, latency_ms: int):
        """Update arm on success"""
        arm.alpha += 1.0
        arm.total_selections += 1
        arm.total_successes += 1

        # Update composite rewards
        latency_score = self._normalize_latency(latency_ms)
        arm.avg_latency_score = self._update_moving_average(
            arm.avg_latency_score, latency_score, arm.total_selections
        )
        arm.avg_success_rate = arm.total_successes / arm.total_selections
        arm.updated_at = int(time.time())

    def _update_arm_failure(self, arm: BanditArm, latency_ms: int):
        """Update arm on failure"""
        arm.beta += 1.0
        arm.total_selections += 1
        arm.total_failures += 1

        latency_score = self._normalize_latency(latency_ms)
        arm.avg_latency_score = self._update_moving_average(
            arm.avg_latency_score, latency_score, arm.total_selections
        )
        arm.avg_success_rate = arm.total_successes / arm.total_selections
        arm.updated_at = int(time.time())

    def _normalize_latency(self, latency_ms: float) -> float:
        """Normalize latency to 0-1 score (lower = better)"""
        if latency_ms <= 0:
            return 1.0
        score = 1.0 - (latency_ms / self.state.max_latency_ms)
        return max(0.0, min(1.0, score))

    def _update_moving_average(
        self, current: float, new_value: float, count: int
    ) -> float:
        """Exponential moving average"""
        if count == 1:
            return new_value
        alpha = 2.0 / (count + 1)
        return alpha * new_value + (1 - alpha) * current

    def _get_healthy_arms(self) -> List[BanditArm]:
        """Get arms with circuit breakers closed"""
        return [
            arm
            for arm in self.state.arms
            if not self.circuit_breakers[arm.provider_id].get_is_open()
        ]

    def _get_top_arms_by_mean(self, n: int) -> List[BanditArm]:
        """Get top N arms by expected value"""
        healthy = self._get_healthy_arms()
        healthy.sort(key=lambda a: a.get_mean(), reverse=True)
        return healthy[:n]

    def update_state(self, state: BayesianState):
        """Update state (called when control plane returns)"""
        self.state = state

        # Update circuit breakers for new arms
        for arm in state.arms:
            if arm.provider_id not in self.circuit_breakers:
                self.circuit_breakers[arm.provider_id] = CircuitBreaker(
                    state.circuit_breaker_threshold
                )

    def get_state(self) -> BayesianState:
        """Get current state"""
        return self.state
